File: src/aws/lambdas/incremental_distribution/src/test/java/uk/nhs/nhsx/diagnosiskeydist/apispec/DailyZIPSubmissionPeriodTest.java
package uk.nhs.nhsx.diagnosiskeydist.apispec;

import org.junit.jupiter.api.Test;

import java.time.Instant;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.util.Calendar;
import java.util.Date;
import java.util.List;
import java.util.TimeZone;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

public class DailyZIPSubmissionPeriodTest {

    @Test
    public void testZipPath() {
        Date endDate = utcDate(2020, 7, 20, 0, 0, 0, 0);
        DailyZIPSubmissionPeriod dailyZIPSubmissionPeriod = new DailyZIPSubmissionPeriod(endDate);
        assertEquals("distribution/daily/2020072000.zip", dailyZIPSubmissionPeriod.zipPath());
    }

    @Test
    public void testIsCoveringSubmissionDate() {
        assertFalse(new DailyZIPSubmissionPeriod(utcDate(2020, 07, 4, 0, 0, 0, 0))
                .isCoveringSubmissionDate(utcDate(2020, 07, 2, 23, 59, 59, 999), 0));
        assertTrue(new DailyZIPSubmissionPeriod(utcDate(2020, 07, 4, 0, 0, 0, 0))
                .isCoveringSubmissionDate(utcDate(2020, 07, 3, 0, 0, 0, 0), 0));
        assertTrue(new DailyZIPSubmissionPeriod(utcDate(2020, 07, 4, 0, 0, 0, 0))
                .isCoveringSubmissionDate(utcDate(2020, 07, 3, 23, 59, 59, 999), 0));
        assertFalse(new DailyZIPSubmissionPeriod(utcDate(2020, 07, 4, 0, 0, 0, 0))
                .isCoveringSubmissionDate(utcDate(2020, 07, 4, 0, 0, 0, 0), 0));

        assertTrue(new DailyZIPSubmissionPeriod(utcDate(2020, 07, 4, 0, 0, 0, 0))
                .isCoveringSubmissionDate(utcDate(2020, 07, 2, 23, 59, 59, 999), -15));
        assertFalse(new DailyZIPSubmissionPeriod(utcDate(2020, 07, 4, 0, 0, 0, 0))
                .isCoveringSubmissionDate(utcDate(2020, 07, 3, 23, 59, 59, 999), -15));
    }

    @Test
    public void testPeriodForSubmissionDate() {
        assertEquals("distribution/daily/2020070500.zip", DailyZIPSubmissionPeriod.periodForSubmissionDate(utcDate(2020, 07, 4, 0, 0, 0, 0)).zipPath());
        assertEquals("distribution/daily/2020070500.zip", DailyZIPSubmissionPeriod.periodForSubmissionDate(utcDate(2020, 07, 4, 23, 59, 59, 999)).zipPath());
    }

    @Test
    public void testAllPeriodsToGenerate() {
        Date endDate = utcDate(2020, 7, 20, 0, 0, 0, 0);
        DailyZIPSubmissionPeriod dailyZIPSubmissionPeriod = new DailyZIPSubmissionPeriod(endDate);

        List<DailyZIPSubmissionPeriod> result = dailyZIPSubmissionPeriod.allPeriodsToGenerate();
        assertEquals(15, result.size());

        //Special case: zip "unofficially" available during the day and contains the keys submitted so far (the mobile apps may only download the ZIP after the end of the day):
        assertEquals("distribution/daily/2020072000.zip", result.get(0).zipPath());

        //Regular cases (API spec)
        assertEquals("distribution/daily/2020071900.zip", result.get(1).zipPath());
        assertEquals("distribution/daily/2020070600.zip", result.get(14).zipPath());
    }

    private int getDayOfMonth(Date dateToConvert) {
        LocalDateTime dateTime = Instant.ofEpochMilli(dateToConvert.getTime())
                .atZone(ZoneId.systemDefault())
                .toLocalDateTime();

        return dateTime.getDayOfMonth();
    }

    private Date utcDate(int year, int month, int day, int hour, int minute, int second, int millis) {
        Calendar cal = Calendar.getInstance(TimeZone.getTimeZone("UTC"));
        cal.set(Calendar.YEAR, year);
        cal.set(Calendar.MONTH, month - 1);
        cal.set(Calendar.DAY_OF_MONTH, day);
        cal.set(Calendar.HOUR_OF_DAY, hour);
        cal.set(Calendar.MINUTE, minute);
        cal.set(Calendar.SECOND, second);
        cal.set(Calendar.MILLISECOND, millis);

        return cal.getTime();
    }
}

File: src/aws/lambdas/incremental_distribution/src/test/java/uk/nhs/nhsx/diagnosiskeydist/apispec/DailyZIPSubmissionPeriodTest.kt
package uk.nhs.nhsx.diagnosiskeydist.apispec

import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.CsvSource
import org.junit.jupiter.params.provider.ValueSource
import uk.nhs.nhsx.testhelper.data.asInstant
import java.time.Duration

class DailyZIPSubmissionPeriodTest {

    @ParameterizedTest
    @CsvSource(
        "2020-07-20T00:00:00.000Z,false",
        "2020-07-20T05:00:00.000Z,true",
        "2020-07-20T00:05:00.000Z,true",
        "2020-07-20T00:00:05.000Z,true",
        "2020-07-20T00:00:00.005Z,true",
        "2020-07-20T05:05:05.005Z,true"
    )
    fun `throws exception if submission period is invalid`(periodEndDate: String, throwsEx: Boolean) {
        val endDate = periodEndDate.asInstant()

        if (throwsEx) {
            assertThatThrownBy { DailyZIPSubmissionPeriod(endDate) }.isInstanceOf(IllegalStateException::class.java)
        } else {
            assertThat(DailyZIPSubmissionPeriod(endDate)).isNotNull
        }
    }

    @ParameterizedTest
    @CsvSource(
        "2020-07-20T00:00:00.000Z,distribution/daily/2020072000.zip"
    )
    fun `creates ZIP path`(periodEndDate: String, expected: String) {
        DailyZIPSubmissionPeriod(periodEndDate.asInstant()).run {
            assertThat(zipPath()).isEqualTo(expected)
        }
    }

    @ParameterizedTest
    @CsvSource(
        "2020-07-04T00:00:00.000Z,2020-07-02T23:59:59.999Z,0,false",
        "2020-07-04T00:00:00.000Z,2020-07-03T00:00:00.000Z,0,true",
        "2020-07-04T00:00:00.000Z,2020-07-03T23:59:59.999Z,0,true",
        "2020-07-04T00:00:00.000Z,2020-07-04T00:00:00.000Z,0,false",
        "2020-07-04T00:00:00.000Z,2020-07-02T23:59:59.999Z,-15,true",
        "2020-07-04T00:00:00.000Z,2020-07-03T23:59:59.999Z,-15,false"
    )
    fun `is covering submission date`(
        periodEndDate: String,
        submissionDate: String,
        offsetMinutes: Long,
        expected: Boolean
    ) {
        DailyZIPSubmissionPeriod(periodEndDate.asInstant())
            .isCoveringSubmissionDate(submissionDate.asInstant(), Duration.ofMinutes(offsetMinutes))
            .run {
                assertThat(this)
                    .describedAs("end date: $periodEndDate submission date: $submissionDate offset: $offsetMinutes")
                    .isEqualTo(expected)
            }
    }

    @ParameterizedTest
    @CsvSource(
        "2020-07-04T00:00:00.000Z,distribution/daily/2020070500.zip",
        "2020-07-04T23:59:59.999Z,distribution/daily/2020070500.zip",

        )
    fun `calculates period for submission date`(submissionDate: String, expected: String) {
        DailyZIPSubmissionPeriod.periodForSubmissionDate(submissionDate.asInstant()).run {
            assertThat(zipPath())
                .describedAs("submission date: $submissionDate")
                .isEqualTo(expected)
        }
    }

    @Test
    fun `verify all periods`() {
        val zipSubmissionPeriod = DailyZIPSubmissionPeriod("2020-07-20T00:00:00.000Z".asInstant())
        val allPeriods = zipSubmissionPeriod.allPeriodsToGenerate()

        assertThat(allPeriods.map { it.zipPath() })
            .hasSize(15)
            .containsExactly(
                // Special case: zip "unofficially" available during the day and
                // contains the keys submitted so far (the mobile apps may only
                // download the ZIP after the end of the day)
                "distribution/daily/2020072000.zip",

                // Regular cases (API spec)
                "distribution/daily/2020071900.zip",
                "distribution/daily/2020071800.zip",
                "distribution/daily/2020071700.zip",
                "distribution/daily/2020071600.zip",
                "distribution/daily/2020071500.zip",
                "distribution/daily/2020071400.zip",
                "distribution/daily/2020071300.zip",
                "distribution/daily/2020071200.zip",
                "distribution/daily/2020071100.zip",
                "distribution/daily/2020071000.zip",
                "distribution/daily/2020070900.zip",
                "distribution/daily/2020070800.zip",
                "distribution/daily/2020070700.zip",
                "distribution/daily/2020070600.zip"
            )
    }

    @ParameterizedTest
    @CsvSource(
        "2020-07-20T00:00:00.000Z,2020-07-19T00:00:00.000Z",
        "2020-07-01T00:00:00.000Z,2020-06-30T00:00:00.000Z",
        "2021-01-01T00:00:00.000Z,2020-12-31T00:00:00.000Z",
    )
    fun `calculate start date inclusive`(periodEndDate: String, expected: String) {
        DailyZIPSubmissionPeriod(periodEndDate.asInstant()).startInclusive.run {
            assertThat(this).isEqualTo(expected.asInstant())
        }
    }

    @ParameterizedTest
    @ValueSource(
        strings = [
            "2020-07-20T00:00:00.000Z",
            "2020-07-19T00:00:00.000Z",
            "2020-07-18T00:00:00.000Z"
        ]
    )
    fun `calculate end date inclusive`(input: String) {
        DailyZIPSubmissionPeriod(input.asInstant()).endExclusive.run {
            assertThat(this).isEqualTo(input.asInstant())
        }
    }

    @Test
    fun `verify toString`() {
        assertThat(DailyZIPSubmissionPeriod("2020-07-20T00:00:00.000Z".asInstant()).toString())
            .isEqualTo("1 day: from 2020071900 (inclusive) to 2020072000 (exclusive)")
    }
}
